jgt_common/__init__.py

```diff
@@ -872,6 +872,101 @@ def check_until(
     raise IncompleteAtTimeoutException(msg, call_result=result, timeout=timeout)
 
 
+@classify("misc", "exceptions")
+def assert_if_values(format_if_format, error_fun=lambda x: "\n".join(truths_from(x))):
+    """
+    Assert if any truthy values are returned (or yielded) from the decorated function.
+
+    Any values that the decorated function yields/returns are passed to ``error_fun``.
+    ``format_if_format``, and the result of the ``error_fun`` call,
+    are passed to ``format_if``.
+    The result of the ``format_if`` call is used for an assertion.
+
+    Args:
+        format_if_format (str): the first parameter to format_if.
+        error_fun (callable): called on the results of the decorated function,
+            return value will be passed as the 2nd parameter to format_if.
+            The default function will only process truthy values.
+            That can be changed by passing in your own custom ``error_fun``
+
+    Returns:
+        None - If no values are returned, or yielded,
+        or the result of ``error_fun`` is falsey, return None.
+
+    Raises:
+        AssertionError: as described above.
+
+    This decorator is meant to simplify code that otherwise has to manually keep
+    track of a list of things to complain about::
+
+        errors = []
+        for thing in things:
+           if bad(thing):
+               errors.append(description_of(thing))
+        err_msg = format_if("blah blah blah: {}", "".join(errors))
+        assert not err_msg, err_msg
+
+    into simpler code::
+
+        @assert_if_values("blah blah blah: {}")
+        ...
+            for thing in things:
+                if bad(thing):
+                    yield description_of(thing)
+
+    """
+
+    @_wrapt.decorator
+    def helper(wrapped, instance, args, kwargs):
+        err_msg = format_if(format_if_format, error_fun(wrapped(*args, **kwargs)))
+        assert not err_msg, err_msg
+
+    return helper
+
+
+@classify("misc", "sequence")
+def accumulator_for(fun):
+    """
+    Accumulate the results of calling fun into a unique list, returning that list.
```

Comment:
NB: The `fun` vs. `func` abbreviation still throws me a bit :) I keep wondering why Doug is enjoying everything so much.

Let's go outside and have some error fun!

Reply:
:laughing:

Reply:
Hey, I'm a QE. Errors _are_ fun!
My use of `fun` is a habit that goes way back to older languages.
I'd kinda prefer to use `fn` over `func`, but that probably goes back too far too.
I did check, and while we have `"func"tools` as a module, `map`, `filter`, and even `functools.reduce` call their first argument `function`.
I have a deep level of disinterest in whichever term we use, but it's probably appropriate to have a separate PR for normalizing that terminology across each jgt repo :-)